rnote-ui/src/workspacebrowser/filerow/actions/duplicate.rs

```diff
@@ -0,0 +1,102 @@
+use std::path::PathBuf;
+
+use fs_extra::dir::{CopyOptions, TransitProcessResult};
+use fs_extra::{copy_items_with_progress, TransitProcess};
+use gtk4::prelude::FileExt;
+use gtk4::{gio, glib, glib::clone};
+
+use crate::workspacebrowser::FileRow;
+
+const DUPLICATE_SUFFIX: &str = ".dup";
+
+impl FileRow {
+    pub fn duplicate_action(&self) -> gio::SimpleAction {
+        let action = gio::SimpleAction::new("duplicate", None);
+
+        action.connect_activate(
+            clone!(@weak self as filerow => move |_action_duplicate_file, _| {
+
+                let yeet = |process: TransitProcess| -> TransitProcessResult {
+                    let status = {
+                        let status = process.copied_bytes / process.total_bytes;
+                        status as f64
+                    };
+                    filerow.canvas_progress().set_fraction(status);
+                    TransitProcessResult::ContinueOrAbort
+                };
```

Comment:
Ok, so I tried to define the progress function (it's this `yeet` closure) to call the progress-canvas, but I'm getting the following error, if I run this:
```rust
 INFO  rnote > ... env_logger initialized
 ERROR rnote_engine::engine > failed to create a new audio player in PenHolder::default(), Err A backend-specific error has occurred: ALSA function 'snd_pcm_open' failed with error 'EHOSTDOWN: Host is down'
 DEBUG selectors::matching  > Matching complex selector :not(:root)svg for svg id=None
 DEBUG selectors::matching  > Matching complex selector :root for svg id=None
 DEBUG selectors::matching  > Matching complex selector image for svg id=None
 DEBUG selectors::matching  > Matching complex selector marker for svg id=None
 DEBUG selectors::matching  > Matching complex selector pattern for svg id=None
 DEBUG selectors::matching  > Matching complex selector symbol for svg id=None
 DEBUG selectors::matching  > Matching complex selector defs for svg id=None
 DEBUG selectors::matching  > Matching complex selector clipPath for svg id=None
 DEBUG selectors::matching  > Matching complex selector mask for svg id=None
 DEBUG selectors::matching  > Matching complex selector marker for svg id=None
 DEBUG selectors::matching  > Matching complex selector desc for svg id=None
 DEBUG selectors::matching  > Matching complex selector title for svg id=None
 DEBUG selectors::matching  > Matching complex selector metadata for svg id=None
 DEBUG selectors::matching  > Matching complex selector pattern for svg id=None
 DEBUG selectors::matching  > Matching complex selector linearGradient for svg id=None
 DEBUG selectors::matching  > Matching complex selector radialGradient for svg id=None
 DEBUG selectors::matching  > Matching complex selector script for svg id=None
 DEBUG selectors::matching  > Matching complex selector style for svg id=None
 DEBUG selectors::matching  > Matching complex selector symbol for svg id=None
 DEBUG selectors::matching  > Matching complex selector :host(use) > symbol for svg id=None
 DEBUG selectors::matching  > Matching complex selector :not(:root)svg for g id=None
 DEBUG selectors::matching  > Matching complex selector image for g id=None
 DEBUG selectors::matching  > Matching complex selector marker for g id=None
 DEBUG selectors::matching  > Matching complex selector pattern for g id=None
 DEBUG selectors::matching  > Matching complex selector symbol for g id=None
 DEBUG selectors::matching  > Matching complex selector defs for g id=None
 DEBUG selectors::matching  > Matching complex selector clipPath for g id=None
 DEBUG selectors::matching  > Matching complex selector mask for g id=None
 DEBUG selectors::matching  > Matching complex selector marker for g id=None
 DEBUG selectors::matching  > Matching complex selector desc for g id=None
 DEBUG selectors::matching  > Matching complex selector title for g id=None
 DEBUG selectors::matching  > Matching complex selector metadata for g id=None
 DEBUG selectors::matching  > Matching complex selector pattern for g id=None
 DEBUG selectors::matching  > Matching complex selector linearGradient for g id=None
 DEBUG selectors::matching  > Matching complex selector radialGradient for g id=None
 DEBUG selectors::matching  > Matching complex selector script for g id=None
 DEBUG selectors::matching  > Matching complex selector style for g id=None
 DEBUG selectors::matching  > Matching complex selector symbol for g id=None
 DEBUG selectors::matching  > Matching complex selector :host(use) > symbol for g id=None
 DEBUG selectors::matching  > Matching complex selector :not(:root)svg for rect id=None
 DEBUG selectors::matching  > Matching complex selector image for rect id=None
 DEBUG selectors::matching  > Matching complex selector marker for rect id=None
 DEBUG selectors::matching  > Matching complex selector pattern for rect id=None
 DEBUG selectors::matching  > Matching complex selector symbol for rect id=None
 DEBUG selectors::matching  > Matching complex selector defs for rect id=None
 DEBUG selectors::matching  > Matching complex selector clipPath for rect id=None
 DEBUG selectors::matching  > Matching complex selector mask for rect id=None
 DEBUG selectors::matching  > Matching complex selector marker for rect id=None
 DEBUG selectors::matching  > Matching complex selector desc for rect id=None
 DEBUG selectors::matching  > Matching complex selector title for rect id=None
 DEBUG selectors::matching  > Matching complex selector metadata for rect id=None
 DEBUG selectors::matching  > Matching complex selector pattern for rect id=None
 DEBUG selectors::matching  > Matching complex selector linearGradient for rect id=None
 DEBUG selectors::matching  > Matching complex selector radialGradient for rect id=None
 DEBUG selectors::matching  > Matching complex selector script for rect id=None
 DEBUG selectors::matching  > Matching complex selector style for rect id=None
 DEBUG selectors::matching  > Matching complex selector symbol for rect id=None
 DEBUG selectors::matching  > Matching complex selector :host(use) > symbol for rect id=None
 DEBUG selectors::matching  > Matching complex selector :not(:root)svg for svg id=None
 DEBUG selectors::matching  > Matching complex selector :root for svg id=None
 DEBUG selectors::matching  > Matching complex selector image for svg id=None
 DEBUG selectors::matching  > Matching complex selector marker for svg id=None
 DEBUG selectors::matching  > Matching complex selector pattern for svg id=None
 DEBUG selectors::matching  > Matching complex selector symbol for svg id=None
 DEBUG selectors::matching  > Matching complex selector defs for svg id=None
 DEBUG selectors::matching  > Matching complex selector clipPath for svg id=None
 DEBUG selectors::matching  > Matching complex selector mask for svg id=None
 DEBUG selectors::matching  > Matching complex selector marker for svg id=None
 DEBUG selectors::matching  > Matching complex selector desc for svg id=None
 DEBUG selectors::matching  > Matching complex selector title for svg id=None
 DEBUG selectors::matching  > Matching complex selector metadata for svg id=None
 DEBUG selectors::matching  > Matching complex selector pattern for svg id=None
 DEBUG selectors::matching  > Matching complex selector linearGradient for svg id=None
 DEBUG selectors::matching  > Matching complex selector radialGradient for svg id=None
 DEBUG selectors::matching  > Matching complex selector script for svg id=None
 DEBUG selectors::matching  > Matching complex selector style for svg id=None
 DEBUG selectors::matching  > Matching complex selector symbol for svg id=None
 DEBUG selectors::matching  > Matching complex selector :host(use) > symbol for svg id=None
 DEBUG selectors::matching  > Matching complex selector :not(:root)svg for g id=None
 DEBUG selectors::matching  > Matching complex selector image for g id=None
 DEBUG selectors::matching  > Matching complex selector marker for g id=None
 DEBUG selectors::matching  > Matching complex selector pattern for g id=None
 DEBUG selectors::matching  > Matching complex selector symbol for g id=None
 DEBUG selectors::matching  > Matching complex selector defs for g id=None
 DEBUG selectors::matching  > Matching complex selector clipPath for g id=None
 DEBUG selectors::matching  > Matching complex selector mask for g id=None
 DEBUG selectors::matching  > Matching complex selector marker for g id=None
 DEBUG selectors::matching  > Matching complex selector desc for g id=None
 DEBUG selectors::matching  > Matching complex selector title for g id=None
 DEBUG selectors::matching  > Matching complex selector metadata for g id=None
 DEBUG selectors::matching  > Matching complex selector pattern for g id=None
 DEBUG selectors::matching  > Matching complex selector linearGradient for g id=None
 DEBUG selectors::matching  > Matching complex selector radialGradient for g id=None
 DEBUG selectors::matching  > Matching complex selector script for g id=None
 DEBUG selectors::matching  > Matching complex selector style for g id=None
 DEBUG selectors::matching  > Matching complex selector symbol for g id=None
 DEBUG selectors::matching  > Matching complex selector :host(use) > symbol for g id=None
 DEBUG selectors::matching  > Matching complex selector :not(:root)svg for rect id=None
 DEBUG selectors::matching  > Matching complex selector image for rect id=None
 DEBUG selectors::matching  > Matching complex selector marker for rect id=None
 DEBUG selectors::matching  > Matching complex selector pattern for rect id=None
 DEBUG selectors::matching  > Matching complex selector symbol for rect id=None
 DEBUG selectors::matching  > Matching complex selector defs for rect id=None
 DEBUG selectors::matching  > Matching complex selector clipPath for rect id=None
 DEBUG selectors::matching  > Matching complex selector mask for rect id=None
 DEBUG selectors::matching  > Matching complex selector marker for rect id=None
 DEBUG selectors::matching  > Matching complex selector desc for rect id=None
 DEBUG selectors::matching  > Matching complex selector title for rect id=None
 DEBUG selectors::matching  > Matching complex selector metadata for rect id=None
 DEBUG selectors::matching  > Matching complex selector pattern for rect id=None
 DEBUG selectors::matching  > Matching complex selector linearGradient for rect id=None
 DEBUG selectors::matching  > Matching complex selector radialGradient for rect id=None
 DEBUG selectors::matching  > Matching complex selector script for rect id=None
 DEBUG selectors::matching  > Matching complex selector style for rect id=None
 DEBUG selectors::matching  > Matching complex selector symbol for rect id=None
 DEBUG selectors::matching  > Matching complex selector :host(use) > symbol for rect id=None
 DEBUG selectors::matching  > Matching complex selector :not(:root)svg for svg id=None
 DEBUG selectors::matching  > Matching complex selector :root for svg id=None
 DEBUG selectors::matching  > Matching complex selector image for svg id=None
 DEBUG selectors::matching  > Matching complex selector marker for svg id=None
 DEBUG selectors::matching  > Matching complex selector pattern for svg id=None
 DEBUG selectors::matching  > Matching complex selector symbol for svg id=None
 DEBUG selectors::matching  > Matching complex selector defs for svg id=None
 DEBUG selectors::matching  > Matching complex selector clipPath for svg id=None
 DEBUG selectors::matching  > Matching complex selector mask for svg id=None
 DEBUG selectors::matching  > Matching complex selector marker for svg id=None
 DEBUG selectors::matching  > Matching complex selector desc for svg id=None
 DEBUG selectors::matching  > Matching complex selector title for svg id=None
 DEBUG selectors::matching  > Matching complex selector metadata for svg id=None
 DEBUG selectors::matching  > Matching complex selector pattern for svg id=None
 DEBUG selectors::matching  > Matching complex selector linearGradient for svg id=None
 DEBUG selectors::matching  > Matching complex selector radialGradient for svg id=None
 DEBUG selectors::matching  > Matching complex selector script for svg id=None
 DEBUG selectors::matching  > Matching complex selector style for svg id=None
 DEBUG selectors::matching  > Matching complex selector symbol for svg id=None
 DEBUG selectors::matching  > Matching complex selector :host(use) > symbol for svg id=None
 DEBUG selectors::matching  > Matching complex selector :not(:root)svg for g id=None
 DEBUG selectors::matching  > Matching complex selector image for g id=None
 DEBUG selectors::matching  > Matching complex selector marker for g id=None
 DEBUG selectors::matching  > Matching complex selector pattern for g id=None
 DEBUG selectors::matching  > Matching complex selector symbol for g id=None
 DEBUG selectors::matching  > Matching complex selector defs for g id=None
 DEBUG selectors::matching  > Matching complex selector clipPath for g id=None
 DEBUG selectors::matching  > Matching complex selector mask for g id=None
 DEBUG selectors::matching  > Matching complex selector marker for g id=None
 DEBUG selectors::matching  > Matching complex selector desc for g id=None
 DEBUG selectors::matching  > Matching complex selector title for g id=None
 DEBUG selectors::matching  > Matching complex selector metadata for g id=None
 DEBUG selectors::matching  > Matching complex selector pattern for g id=None
 DEBUG selectors::matching  > Matching complex selector linearGradient for g id=None
 DEBUG selectors::matching  > Matching complex selector radialGradient for g id=None
 DEBUG selectors::matching  > Matching complex selector script for g id=None
 DEBUG selectors::matching  > Matching complex selector style for g id=None
 DEBUG selectors::matching  > Matching complex selector symbol for g id=None
 DEBUG selectors::matching  > Matching complex selector :host(use) > symbol for g id=None
 DEBUG selectors::matching  > Matching complex selector :not(:root)svg for rect id=None
 DEBUG selectors::matching  > Matching complex selector image for rect id=None
 DEBUG selectors::matching  > Matching complex selector marker for rect id=None
 DEBUG selectors::matching  > Matching complex selector pattern for rect id=None
 DEBUG selectors::matching  > Matching complex selector symbol for rect id=None
 DEBUG selectors::matching  > Matching complex selector defs for rect id=None
 DEBUG selectors::matching  > Matching complex selector clipPath for rect id=None
 DEBUG selectors::matching  > Matching complex selector mask for rect id=None
 DEBUG selectors::matching  > Matching complex selector marker for rect id=None
 DEBUG selectors::matching  > Matching complex selector desc for rect id=None
 DEBUG selectors::matching  > Matching complex selector title for rect id=None
 DEBUG selectors::matching  > Matching complex selector metadata for rect id=None
 DEBUG selectors::matching  > Matching complex selector pattern for rect id=None
 DEBUG selectors::matching  > Matching complex selector linearGradient for rect id=None
 DEBUG selectors::matching  > Matching complex selector radialGradient for rect id=None
 DEBUG selectors::matching  > Matching complex selector script for rect id=None
 DEBUG selectors::matching  > Matching complex selector style for rect id=None
 DEBUG selectors::matching  > Matching complex selector symbol for rect id=None
 DEBUG selectors::matching  > Matching complex selector :host(use) > symbol for rect id=None

(rnote:27512): Gtk-CRITICAL **: 12:53:25.752: Unable to retrieve child object 'canvas_progress' from class template for type 'FileRow' while building a 'FileRow'
thread 'main' panicked at 'Failed to retrieve template child. Please check that it has been bound and has a #[template_child] attribute.', /main/Programming/pull_requests/rnote/build/cargo-home/registry/src/github.com-1ecc6299db9ec823/gtk4-0.4.8/src/subclass/widget.rs:1305:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
do you have an idea why this is faililng?

Reply:
@flxzt

Reply:
I am not sure what you are trying to do by adding `canvas_progress` here. That widget does not exist as a child of the `file_row`, but you can get it by calling appwindow.canvas_progress()` Just always pass in `appwindow` to the action constructor methods and add it as a weak clone in the `clone!` macro when you are creating callbacks.

Reply:
> I am not sure what you are trying to do by adding canvas_progress here. 

I just took a look into your `appwindow.rs` implementation because I've absolutely no idea how to work with gtk4 and rust xD